internal/generate/clusterserviceversion/bases/markers.go

```diff
@@ -21,31 +21,32 @@ import (
 
 	"github.com/markbates/inflect"
 	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/operator-framework/operator-registry/pkg/registry"
 	log "github.com/sirupsen/logrus"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 
+	"github.com/operator-framework/operator-sdk/internal/generate/clusterserviceversion/bases/definitions"
 	"github.com/operator-framework/operator-sdk/internal/generate/olm-catalog/descriptor"
 )
 
-// updateDescriptionsForGVKs updates csv with API metadata found in apisDir
-// filtered by gvks.
-func updateDescriptionsForGVKs(csv *v1alpha1.ClusterServiceVersion, apisDir string,
-	gvks []schema.GroupVersionKind) error {
-
-	gvkMap := make(map[schema.GroupVersionKind]v1alpha1.CRDDescription)
-	for _, desc := range csv.Spec.CustomResourceDefinitions.Owned {
-		group := desc.Name
-		if split := strings.Split(desc.Name, "."); len(split) > 1 {
-			group = strings.Join(split[1:], ".")
+// updateDefinitions parses APIs in apisDir for code and markers that can build a crdDescription and
+// updates existing crdDescriptions in csv. If no code/markers are found, the crdDescription is appended as-is.
+func updateDefinitions(csv *v1alpha1.ClusterServiceVersion, apisDir string, gvks []schema.GroupVersionKind) error {
+	keys := make([]registry.DefinitionKey, len(gvks))
+	for i, gvk := range gvks {
+		keys[i] = registry.DefinitionKey{
+			Name:    fmt.Sprintf("%s.%s", inflect.Pluralize(strings.ToLower(gvk.Kind)), gvk.Group),
```

Comment:
This is an issue that cropped up from separating manifest collection from CSV generation, where we don't have the CRDs themselves so we can't know their resource (plural) names. Most of the time this will work, but if a `+kubebuilder:resource` marker exists that changes the resource name then these names will be out-of-sync. The right thing to do is parse this marker, or (easier) pass this value to the CSV generator.

Reply:
Fixed by parsing `+kubebuilder:resource` markers, like `controller-gen` does.